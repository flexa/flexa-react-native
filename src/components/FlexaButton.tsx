import * as React from 'react';
import { TouchableOpacity, Image, StyleSheet, ViewStyle } from 'react-native';
import { payment } from '../index';
import type { AppAccount } from '../initialize';

type FlexaButtonProps = {
  appAccounts: AppAccount[];
  paymentCallback: Function;
  customAuth?: Function;
  width?: number;
  height?: number;
  borderRadius?: number;
  style?: ViewStyle;
};
const FlexaButton = ({
  appAccounts,
  paymentCallback,
  customAuth,
  width = 64,
  height = 64,
  borderRadius = 6,
  style = {},
}: FlexaButtonProps) => {
  const handleClick = async () => {
    customAuth && (await customAuth());
    await payment(appAccounts, paymentCallback);
  };
  return (
    <TouchableOpacity
      style={[styles.button, { width, height }, { ...style }]}
      onPress={handleClick}
    >
      <Image
        style={{
          flex: 1,
          width: width,
          height: height,
          borderRadius: borderRadius,
        }}
        resizeMode="contain"
        source={require('./flexaLogo.png')}
      />
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default FlexaButton;
