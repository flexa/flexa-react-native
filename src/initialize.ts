import { NativeModules, Platform } from 'react-native';

const LINKING_ERROR =
  `The package '@flexa/flexa-react-native' doesn't seem to be linked. Make sure: \n\n` +
  Platform?.select({ ios: "- You have run 'pod install'\n", default: '' }) +
  '- You rebuilt the app after installing the package\n' +
  '- You are not using Expo managed workflow\n';

type AvailableAsset = {
  assetId: string;
  icon?: string;
  displayName: string;
  symbol: string;
  balance: number;
  balanceAvailable?: number;
};

const defaultThemingData = {
  android: {
    light: {},
    dark: {},
  },
  iOS: {
    light: {},
    dark: {},
  },
};

export enum CUSTODY_MODEL {
  LOCAL = 'LOCAL',
  MANAGED = 'MANAGED',
}

type CustodyModelType = CUSTODY_MODEL.LOCAL | CUSTODY_MODEL.MANAGED | string;

export type AssetAccount = {
  assetAccountHash: string;
  displayName: string;
  custodyModel: CustodyModelType;
  icon?: string;
  availableAssets: AvailableAsset[];
};

type Transaction = {
  commerceSessionId: string;
  destinationAddress: string;
  amount: string;
  feePriorityPrice: string;
  feePrice: string;
  size: string;
  assetId: string;
  assetAccountHash: string;
};

export type TransactionRequest = {
  transaction: Transaction;
  transactionSent: Function;
};

export const getNativeFlexaModule = () =>
  NativeModules.FlexaReactNative
    ? NativeModules.FlexaReactNative
    : new Proxy(
      {},
      {
        get() {
          throw new Error(LINKING_ERROR);
        },
      }
    );

/**
 * Opens the Flexa for Flexa Payments
 * Pass an assetAccounts array as a parameter and a payment callback function
 *
 * @example
 * try {
 *   await payment(assetAccounts, onPaymentCallback)
 * } catch (e) {
 *   console.error(e)
 * }
 */
export function payment(
  assetAccounts: AssetAccount[],
  onPaymentCallback: Function
): Promise<object> {
  return new Promise((resolve) => {
    const transactionRequest = async (transaction: Transaction) => {
      await onPaymentCallback({
        transaction,
        transactionSent: (txSignature: string) =>
          getNativeFlexaModule().transactionSent(
            transaction.commerceSessionId,
            txSignature
          ),
      });
      getNativeFlexaModule().updatePaymentCallback(assetAccounts, transactionRequest, onPaymentCallback)
    }
    getNativeFlexaModule().payment(
      assetAccounts,
      transactionRequest,
      onPaymentCallback
    );
    resolve({});
  });
}

/**
 * Initializes the Flexa SDK module, please run this on the top layer of the app
 *
 * @example
 *
 * init(
 *  publishableKey: string,
 *  assetAccounts: AssetAccount[],
 *  webViewThemingData: object
 * )
 *
 */
export function init(
  publishableKey: string,
  assetAccounts: AssetAccount[] = [],
  webViewThemingData?: object
): void {
  const themingDataString = JSON.stringify(
    webViewThemingData || defaultThemingData
  );
  const fn = Platform.select({
    ios: () => {
      const iosThemingData = {
        webViewThemeConfig: webViewThemingData || defaultThemingData,
      };
      getNativeFlexaModule().init(
        publishableKey,
        assetAccounts,
        JSON.stringify(iosThemingData)
      );
    },
    android: () => {
      getNativeFlexaModule().init(
        publishableKey,
        assetAccounts,
        themingDataString
      );
    },
  });

  fn?.();
}

export function processUniversalLink(url: string) {
  const fn = Platform.select({
    ios: () =>
      getNativeFlexaModule().processUniversalLink(url, (result: any) => {
        console.log('RNFlexa.processUniversalLink: callback called: ', result);
      }),
    android: () => {
      getNativeFlexaModule().processUniversalLink(url);
    },
  });

  fn?.();
}

export function dismissAllModals() {
  const fn = Platform.select({
    ios: () =>
      new Promise((resolve, reject) => {
        try {
          getNativeFlexaModule().dismissAllModals((result: any) => {
            resolve(result);
          })
        } catch (e) {
          reject(e)
        }
      }),
    android: () =>
      new Promise((resolve, reject) => {
        try {
          getNativeFlexaModule().dismissAllModals((result: any) => {
            resolve(result);
          })
        } catch (e) {
          reject(e)
        }
      })
  });

  return fn?.();
}

/**
 * Updates the Flexa SDK assetAccounts with new availableAssets and balances
 * Pass an assetAccounts array as a parameter
 *
 * @example
 * try {
 *   updateAssetAccounts(assetAccounts)
 * } catch (e) {
 *   console.error(e)
 * }
 */
export function updateAssetAccounts(
  assetAccounts: AssetAccount[],
): void {
  const fn = Platform.select({
    ios: () => {
      getNativeFlexaModule().updateAssetAccounts(
        assetAccounts,
      );
    },
    android: () => {
      getNativeFlexaModule().updateAssetAccounts(
        assetAccounts,
      );
    },
  });

  fn?.();
}
