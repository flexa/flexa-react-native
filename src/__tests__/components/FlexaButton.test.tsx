import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import * as Flexa from '../../initialize'
import FlexaButton from '../../components/FlexaButton';

describe('FlexaButton tests', () => {
  it('matches snapshot.',  () => {
    const paymentSpy = jest.fn();
    // @ts-ignore
    Flexa.payment = paymentSpy;
    const paymentCallback = () => null
    const result = render(
      <FlexaButton assetAccounts={[]} paymentCallback={paymentCallback} />
    );

    expect(result).toMatchSnapshot();
  });

  it('initiates a payment call after being pressed.',  () => {
    const paymentSpy = jest.fn();
    // @ts-ignore
    Flexa.payment = paymentSpy;
    const paymentCallback = () => null
    const result = render(
      <FlexaButton assetAccounts={[]} paymentCallback={paymentCallback} />
    );

    fireEvent(result.getByTestId('FlexaButtonTouchableOpacity'), 'press');

    expect(result).not.toBeNull();
    expect(paymentSpy).toHaveBeenCalled();
    expect(paymentSpy).toHaveBeenCalledWith([], paymentCallback);
  });
});
