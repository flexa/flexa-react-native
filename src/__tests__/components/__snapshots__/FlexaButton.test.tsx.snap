// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexaButton tests matches snapshot. 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "height": 64,
      "justifyContent": "center",
      "opacity": 1,
      "width": 64,
    }
  }
  testID="FlexaButtonTouchableOpacity"
>
  <Image
    resizeMode="contain"
    source={
      {
        "testUri": "../../../src/components/flexaLogo.png",
      }
    }
    style={
      {
        "borderRadius": 6,
        "flex": 1,
        "height": 64,
        "width": 64,
      }
    }
  />
</View>
`;
